### DISCUSSION TITLE
What event should I be using? Moving person detection via Node-RED

### DISCUSSION THREAD

**Discussion by szaske** [blakeblackshear/frigate#2940] [Ask A Question]
I'm not sure if Frigate can do this or not, but I'm hoping someone can help me understand it better.

Using Node-Red I have Frigate alerting me via Alexa when a person is detected in my driveway zone. So now, when someone walks into my driveway my Alexa says "A person is detected in the driveway". Works great....except...when my wife is out in the driveway doing work. in that case it's the same person walking around in the driveway. I added a limiter to my NR flow so that the state could only change every 30 seconds, but even with that added, every 30 seconds Alexa says "A person was detected" when my wife it in the driveway doing chores.  I have checked the state change node option that is "Ignore when current state equals previous state", but still The node will fire every 30 seconds. I was hoping, perhaps with the new "Stationary" functionality that Frigate my have some other event that I can use instead of state change on the Driveway-Zone Person entity so that Frigate will detect when one person enters my driveway...and will be able to track that person the whole time while they stay in the driveway-zone, and then the state will only change when the person leaves the zone. Does Frigate have this ability?  As it works now, if I watch it live (at 5 fps detect btw)...Frigate will see the person and bound-box them...and my Node-RED flow will fire and alert me via Alexa...and then as the person walks within that zone, the boundingbox will disappear, leading me to believe Frigate lost the person...and then a second later it will re-connect with the person and fire another "state change" event (AKA detect a person in the zone).  

So it really looks like Frigate, as I have it configured is not able to track a person over a long period of time (say 15 minutes) while moving around within a zone. Is there another way I can achieve this? Or is there another state-machine-like event I could use instead?

**Comment by NickM-27** [comment]
So a couple things here. 

1. Stationary tracking is for truly stationary (as in a car parked in the driveway) and not staying within a zone. 
2. I am not very familiar with Node-Red so I am not 100% sure what the "ignore when current state equals previous state" but that sounds to me like it is matching all MQTT vars. Have you tried just limiting the event to run when the Frigate event has `type == new' because otherwise the type will be 'updated' or 'end'. With the new type it will only trigger when a person is first seen. 
3. If the above is not working then you may want to update the `max_disappeared` in detect for that camera. By default frigate will consider an object gone if it isn't detected within 25 frames (5 seconds if default 5fps is set). Increasing this will mean if your wife goes out of frame and comes back in the same event will be updated instead of creating a new one. This would be a last resort as it does have some side effects like events taking longer to end and potential for another person to be mistaken as the same one as earlier.

**Comment by szaske** [comment]
1. That's what I assumed about "Stationary"
2. Node Red state node will fire anytime the state equals what you have defined. In this case I'm using an entity called sensor.driveway_zone_person...which is a binary sensor of sorts that can output either true or false. When the sensor sees TRUE to starts my flow (a flow is like a script in Node-RED).  The checkbox labeled "(Ignore) Current state equals previous state", means that the node will not fire a 2nd or 3rd time if the current state is TRUE...and another TRUE state occurs it just ignores it instead. So it's a clumsy way of attempting a state machine...that is the node will only fire when the state goes from FALSE to TRUE. There is another node type called Trigger State Node that allows me to add conditions to the state event. I'll see if your suggestion of type==new will help. It should, at least a little bit....but what I really need if for Frigate to see a person enter a zone and then sort of assign an identity to that person...so that Frigate was smart enough to understand that Person A has entered the zone. And if person A walks behind a bush and then reappears Frigate needs to understand that this is the same Person A and is not a new detection of Person B. And this is all the same one event, Person A entering and moving around in the zone.  That would be perfect. Frigate attaches the person ID to the event, so I could look for a person detection that includes checks to see if the last person detection ID is equal to the current person ID...and if they're the same then I could ignore those events. So then my Node-RED script would look something like this.

A. Alert me when a NEW person detection event occurs AND when the person ID is not equal to the last detection person ID.
B. I'd probably want this to time out after a certain number of minutes say 4 minutes.

So my wife enters the driveway and starts to wash the car...Node Red alerts me and identifies my wife as Person A. my wife walks around the car and is identified 12 more times...but Frigate always understands that this is the same person...Person A. so no new Node Red events fire.

My wife leaves the driveway to make a phone call and returns after 10 minutes...in which case since it's been longer than 4 minutes Node-Red alerts me that a person has been detected.

A second person appears while my wife is washing the car, Frigate determines this is Person B and alerts me that a person has entered the driveway.

I understand Frigate most likely does not have this functionality...but it's something I'd love to see Frigate work towards. Maybe this is how that type == new meta data is determined?

**Reply by NickM-27** [reply]
I mean frigate doesn't know people but it definitely tracks people separately where they each have their own ID until it is stopped being detected. I think your implementation could certainly be a lot better if the MQTT topic was used instead of the home assistant sensor which provides much less contextual information

**Reply by szaske** [reply]
I agree I'll look into switching to MQTT and see if that helps, thanks for the responses.

**Reply by NickM-27** [reply]
You could also use double-take which does facial recognition on top of frigate and key off of that

**Comment by blakeblackshear** [comment]
The [MQTT events topic](https://docs.frigate.video/integrations/mqtt#frigateevents) should have everything you need here. Think of it like a change feed so you can customize when you want notifications based on the before and after values to only notify when things change. For example, you could only notify when the entered_zones contains the zone of interest in the after but not in the before which would mean that you only get notifications the first time an object enters the zone.
